### PR TITLE
Pre-signed urls for map submission

### LINKED ISSUE #318: Chunked Map Uploading for Maps larger than 100MB
Our file uploading process is currently just a single POST request to the backend. It'd be great to get resume-able uploads with multi part chunking it, especially when we move to cloud storage.

The frontend should update visually to respect this!

### LINKED ISSUE #705: S3 Pre-signed URLs for upload endpoints
We want to keep traffic through the actual API to a minimum. There shouldn't be any need to stream data like map uploads through the API itself, instead, just generate pre-signed S3 URLs and pass those to the client.

We should go through every endpoint that does anything upload/download related and ensure that where possible, we avoid file streams. The game seems like it's already using it for map downloads, not sure about images and runs though - I'll look at that myself.

### PR DESCRIPTION
**Pull request by GordiNoki**
Closes #705
Closes #318 

When uploading BSP for map submission s3 pre-signed urls will be used now.

### Checks

- [x] I have included/updated tests where applicable (see [Testing](https://github.com/momentum-mod/website/wiki/Testing))
- [x] I have ran `nx run db:create-migration` and committed the migration if I've made DB schema changes
- [x] I have followed [semantic commit messages](https://gist.github.com/joshbuchea/6f47e86d2510bce28f8e7f42ae84c716) e.g. `feat: Add foo`, `chore: Update bar`, etc...
- [x] My branch has a clear history of changes that can be easy to follow when being reviewed commit-by-commit
- [x] My branch is functionally complete; the only changes to be done will be those potentially requested in code review
- [x] All changes requested in review have been `fixup`ed into my original commits

### CODE REVIEW

File: apps/backend/src/app/dto/map/map-pre-signed-url.dto.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import { MapPreSignedUrl } from '@momentum/constants';
+import { ApiProperty } from '@nestjs/swagger';
+import { IsString, IsUrl } from 'class-validator';
+
+export class MapPreSignedUrlDto implements MapPreSignedUrl {
+  @ApiProperty()
+  @IsString()
```
Comment (tsa96) [inline apps/backend/src/app/dto/map/map-pre-signed-url.dto.ts:7]: Redundant, `@IsUrl` will check that this is a string.

File: apps/backend/src/app/modules/maps/maps.controller.ts [hunk 2]
```diff
@@ -189,6 +190,18 @@ export class MapsController {
     return this.mapsService.getAll(query, userID);
   }
 
+  @Get('/preSignedUrl')
```
Comment (tsa96) [inline apps/backend/src/app/modules/maps/maps.controller.ts:193]: Weird name, let's do `getMapUploadUrl`. I know that's a bit pedantic but should be an easy find-and-replace.

File: apps/backend/src/app/modules/maps/maps.controller.ts [hunk 3]
```diff
@@ -189,6 +190,18 @@ export class MapsController {
...
+  @UseGuards(KillswitchGuard)
+  @Killswitch(KillswitchType.MAP_SUBMISSION)
+  @ApiOperation({ summary: 'Get pre-sigined url for big map uploads' })
```
Comment (tsa96) [inline apps/backend/src/app/modules/maps/maps.controller.ts:196]: Typo, also this is for all map uploads.

File: apps/backend/src/app/modules/maps/maps.service.ts [hunk 4]
```diff
@@ -1843,6 +1883,27 @@ export class MapsService {
...
+  async getBspFromTemp(userID: number): Promise<File> {
+    const userIDObjects = await this.fileStoreService.listFileKeys(
+      `upload_tmp/${userID}`
+    );
+
+    if (userIDObjects.length === 0) {
+      throw new BadRequestException('Missing BSP file');
+    }
+
+    const headerData = await this.fileStoreService.getFile(userIDObjects[0]);
```
Comment (tsa96) [inline apps/backend/src/app/modules/maps/maps.service.ts:1895]: Weird variable name - isn't this the whole buffer?

File: apps/backend-e2e/src/maps.e2e-spec.ts [hunk 5]
```diff
@@ -972,25 +988,15 @@ describe('Maps', () => {
...
-            status: 204,
-            body: {
-              NEW_SIGNUPS: false,
-              RUN_SUBMISSION: false,
-              MAP_SUBMISSION: false,
-              MAP_REVIEWS: false
-            },
-            token: adminToken
-          });
+          await resetKillswitches(req, adminToken);
```
Comment (tsa96) [inline apps/backend-e2e/src/maps.e2e-spec.ts:999]: In future, this should be a separate commit.

**Review by tsa96 — CHANGES_REQUESTED** [review]
Really great work! Though, the `mapSubmissionVersion` endpoint has essentially the behavior as this and also needs changing.

File: apps/backend/src/app/modules/maps/maps.service.ts (hunk 4 above)
Comment (GordiNoki) [reply apps/backend/src/app/modules/maps/maps.service.ts:1895]: It was just a header before

**Comment by GordiNoki** [comment]
> Though, the `mapSubmissionVersion` endpoint has essentially the behavior as this and also needs changing.

I did do the same behavior for POST /maps/:mapID

**Comment by GordiNoki** [comment]
I do really need to figure out why eslint doesn't work in my ide...

**Comment by tsa96** [comment]
> > Though, the `mapSubmissionVersion` endpoint has essentially the behavior as this and also needs changing.
> 
> I did do the same behavior for POST /maps/:mapID

Oh right, I'm an idiot. Aren't you missing frontend stuff for it though?

File: apps/backend/src/app/modules/maps/maps.service.ts (hunk 4 above)
Comment (tsa96) [reply apps/backend/src/app/modules/maps/maps.service.ts]: It's a byte array!

**Review by tsa96 — APPROVED** [review]
Okay LGTM, can merge once CI passes

**Pull request merged**